File: PAT_Basic_Level_by_lua/1051 复数乘法/1051 复数乘法.lua
-- 题目编号: 1051 复数乘法
-- 实现原理: 读取两个复数的极坐标形式(R,P)，根据复数乘法公式(A*B)的模为R1*R2，辐角为P1+P2，计算直角坐标结果。将绝对值小于0.005的值视为0处理，避免输出-0.00。
-- 读取两个复数的极坐标参数：R1(模), P1(辐角), R2(模), P2(辐角)
local a, b, c, d = io.read("*n"), io.read("*n"), io.read("*n"), io.read("*n")
-- 计算乘积的实部和虚部
-- 复数乘法：模相乘 a*c，辐角相加 b+d
-- 实部 x = 模 * cos(辐角)，虚部 y = 模 * sin(辐角)
local x, y = a * c * math.cos(b + d), a * c * math.sin(b + d)
-- 将绝对值小于0.005的值视为0，避免输出-0.00这样的负零结果
if math.abs(x) < .005 then x = 0 end
if math.abs(y) < .005 then y = 0 end
-- 按格式输出：实部保留两位小数，虚部带正负号保留两位小数，以i结尾
print(string.format("%.2f%+.2fi", x, y))

File: PAT_Basic_Level_by_ruby/1051 复数乘法/1051 复数乘法.rb
# 1051 复数乘法
#
# 题目描述：
#   复数可以写成 (A + Bi) 的常规形式，其中 A 是实部，B 是虚部，i 是虚数单位，满足 i^2 = -1；也可以写成极坐标下的指数形式 (R\times e^{(Pi)})，其中 R 是复数模，P 是辐角，i 是虚数单位，其等价于三角形式 R(\cos (P) + i \sin (P))。
#   现给定两个复数的 R 和 P，要求输出两数乘积的常规形式。
#
# 实现原理：
#   实现复数的乘法运算；使用极坐标形式：模相乘，幅角相加；结果转换为常规形式A+Bi
#

if __FILE__ == $PROGRAM_NAME
  r1, p1, r2, p2 = gets.split.map(&:to_f)
  r = r1 * r2
  p = p1 + p2
  real = r * Math.cos(p)
  imag = r * Math.sin(p)
  real = 0.0 if real.abs < 0.005
  imag = 0.0 if imag.abs < 0.005

  printf "%.2f%+.2fi\n", real, imag
end
